File: Day 09 - Array, Strings and Vector/Longest Mountain in Array.cpp
#include <bits/stdc++.h>
using namespace std;

int longestMountain(vector<int> A)
{
  int N = A.size(), res = 0;
  vector<int> up(N, 0), down(N, 0);
  for (int i = N - 2; i >= 0; --i)
    if (A[i] > A[i + 1])
      down[i] = down[i + 1] + 1;
  for (int i = 0; i < N; ++i)
  {
    if (i > 0 && A[i] > A[i - 1])
      up[i] = up[i - 1] + 1;
    if (up[i] && down[i])
      res = max(res, up[i] + down[i] + 1);
  }
  return res;
}

int main()
{

  return 0;
}

File: Day 09 - Array, Strings and Vector/Longest Mountain in Array.java
class Solution {
    public int longestMountain(int[] arr) {
		int N = arr.length;
		int[] pref = new int[N];
		for (int i = 1; i < N; i++) {
			if (arr[i - 1] < arr[i]) {
				pref[i] = pref[i - 1] + 1;
			}
		}
		int[] suff = new int[N];
		for (int i = N - 2; i >= 0; i--) {
			if (arr[i] > arr[i + 1]) {
				suff[i] = suff[i + 1] + 1;
			}
		}
		int ans = 0;
		for (int i = 1; i < N - 1; i++) {
			int left = pref[i];
			int right = suff[i];
			if (left > 0 && right > 0)
				ans = Math.max(ans, left + right + 1);
		}
		return ans;
	}
}